### PR TITLE
Added Filter Scenarios by time  on CycleManager Class

### LINKED ISSUE #413: Cycle based filtering for scenario getters
Propose a way to filter scenarios by providing a cycle description. 

# Objectives
We want to implement a filter mechanism, a query system, or Python API to easily describe and retrieve a meaningful subset of scenarios. This filtering should be based on the cycle. (thanks to @Forchapeatl Pull Request 😉, I realize that even if we want to expose an API to retrieve a subset of scenarios, the implementation is related to cycle description).

# use cases:

In the following, I do not propose implementing a `get_scenarios` function exactly as written. It is the easiest way I found to describe what should be provided by the user and what should be returned by the filtering mechanism. 
These are some use cases I have in mind. 
Similar descriptions for weekly, monthly, quarterly, and yearly scenarios should be provided as an output of this issue. Once we have listed all the relevant use cases. An API implementation can be started.

### Daily scenarios
Here are a few use cases I have in mind for daily scenarios. 

- `get_scenarios(days_of_week="week_days")`
Returns daily scenarios whose cycle corresponds to a weekday
- `get_scenarios(days_of_week="week_end")`
Returns daily scenarios whose cycle corresponds to a Saturday or a Sunday
- `get_scenarios(days_of_week=["Mondays", "Thursdays"])`
Returns daily scenarios whose cycle corresponds to a Monday or a Thursday
- `get_scenarios(days_of_month=[1, 15, 21])`
Returns daily scenarios whose cycle corresponds to the 1st, the 15th, or the 21st day of a month.
- `get_scenarios(days_of_month="last")`
Returns daily scenarios whose cycle corresponds to the last day of the month (28, 29, 30, or 31, depending on the month).
- `get_scenarios(days_of_year=["1/1", "12/5", "22/8"])`
Returns daily scenarios whose cycle corresponds to specific dates: Jan 1st, May 12th, or Aug 22nd.
- `get_scenarios_by(days_of_year=["January", "March"])`
Returns daily scenarios whose cycle corresponds to a January or March day.

more info in comments: https://github.com/Avaiga/taipy/issues/413#issuecomment-2259687098

### PR DESCRIPTION
**Pull request by Forchapeatl**
This PR continues https://github.com/Avaiga/taipy/pull/646

### CODE REVIEW

File: taipy/core/cycle/_cycle_manager.py [hunk 1]
```diff
@@ -150,3 +150,14 @@ def _get_cycles_by_frequency_and_overlapping_date(
...
+    @classmethod
+    def _get_scenarios_by_time(cls , at: datetime):
+        from ..scenario._scenario_manager_factory import _ScenarioManagerFactory
+
+        scenarios = []
+        for scenario in _ScenarioManagerFactory._build_manager()._get_all():
+            cycle = scenario.cycle
+            if cycle.start_date <= at and at <=  cycle.end_date:
+                scenarios.append(scenario)
+        return scenarios
```
Comment (trgiangdo) [inline taipy/core/cycle/_cycle_manager.py]: The method is in CycleManager now, which is great.

However, you are not utilizing any method of this class but still calling to the `_ScenarioManager._get_all()`, which results in a local import.

In the _CycleManager, you can see that we have `_get_cycles_by_frequency_and_overlapping_date()` and `_get_cycles_by_frequency_and_start_date()` methods. Probably you can use this 2 methods, or create something similar.

Then, on `taipy/core/taipy.py`, you can expose a `get_scenarios_by_time` method, which get cycles by the `at` parameter, then get scenario.

File: tests/core/cycle/test_cycle_manager.py [hunk 2]
```diff
@@ -290,3 +290,21 @@ def test_get_primary(tmpdir, cycle, current_datetime):
         )
         == 0
     )
+
+
+def test_get_scenarios_by_time(at=datetime(2024, 1, 17, 1, 0, 0, 0)):
```
Comment (trgiangdo) [inline tests/core/cycle/test_cycle_manager.py]:
```suggestion
def test_get_scenarios_by_time():
    at_time = datetime(2024, 1, 17, 1, 0, 0, 0)
```
The `at` should not be a function parameter

**Review by trgiangdo — COMMENTED** [review]
Thank you @Forchapeatl. Let's continue our discussion to find a way to improve your PR

**Comment by Forchapeatl** [comment]
Hello  @trgiangdo ,  I have been guided by [this comment](https://github.com/Avaiga/taipy/pull/646#issuecomment-1894866390) it helped me a lot, thank you. What worries me is :
1. If I use the `_CycleManager._get_all()` , I will get a list of cycles but I have n't found a means to extract  scenarios form a cycle.
 Even this [get_cycles_scenarios](https://docs.taipy.io/en/latest/manuals/reference/taipy.core.get_cycles_scenarios/#taipy.core.get_cycles_scenarios) method is from `get_scenarios` calls the ` scenarioManager` Class.  I think it would be hard to avoid the local import of the scenario Manager. ( I may be wrong )

2. I am not really sure of my logic. If there is a way to get the scenarios of a cycle, please  help me with hints .

**Comment by trgiangdo** [comment]
You can get the scenarios by cycle using `_ScenarioManager._get_all_by_cycle()` method. But this should not be called inside the _CycleManager, it should be called at the API

File: taipy/core/cycle/_cycle_manager.py [hunk 3]
```diff
@@ -150,3 +150,14 @@ def _get_cycles_by_frequency_and_overlapping_date(
...
         return [cycle for cycle in cycles if cdt(cycle)]
+
+    @classmethod
+    def _get_scenarios_by_time(cls , at: datetime):
+        from ..scenario._scenario_manager_factory import _ScenarioManagerFactory
+
+        scenarios = []
+        for scenario in _ScenarioManagerFactory._build_manager()._get_all():
+            cycle = scenario.cycle
+            if cycle.start_date <= at and at <=  cycle.end_date:
```
Comment (florian-vuillemot) [inline taipy/core/cycle/_cycle_manager.py]:
```suggestion
            if cycle.start_date <= at <= cycle.end_date:
```

File: taipy/core/cycle/_cycle_manager.py (hunk 1 above)
Comment (florian-vuillemot) [inline taipy/core/cycle/_cycle_manager.py]:
```suggestion
        return [
          sc for sc in _ScenarioManagerFactory._build_manager()._get_all()
          if sc.start_date <= at <= sc.end_date
        ]
```

**Comment by Forchapeatl** [comment]
# Example Usage 

```py
from taipy.config import Config, Frequency, Scope
import taipy as tp
import datetime as dt
from datetime import date

import pandas as pd


def filter_by_month(df, month):
    df['Date'] = pd.to_datetime(df['Date']) 
    df = df[df['Date'].dt.month == month]
    return df


historical_data_cfg = Config.configure_csv_data_node(id="historical_data",
                                                     default_path="time_series.csv",
                                                     scope=Scope.GLOBAL)
month_cfg =  Config.configure_data_node(id="month",
                                        scope=Scope.CYCLE)
month_values_cfg =  Config.configure_data_node(id="month_data",
                                               scope=Scope.CYCLE)


task_filter_cfg = Config.configure_task(id="filter_by_month",
                                        function=filter_by_month,
                                        input=[historical_data_cfg, month_cfg],
                                        output=month_values_cfg)


scenario_cfg = Config.configure_scenario(id="my_scenario",
                                         task_configs=[task_filter_cfg],
                                         frequency=Frequency.MONTHLY)


if __name__ == '__main__':
    tp.Core().run()

    scenario_1 = tp.create_scenario(scenario_cfg,
                                    creation_date=dt.datetime(2022,10,7),
                                    name="Scenario 2022/10/7")
    scenario_2 = tp.create_scenario(scenario_cfg,
                                    creation_date=dt.datetime(2022,10,5),
                                    name="Scenario 2022/10/5")

    scenario_1.month.write(10)
    scenario_1.submit()
    cycle = scenario_1.cycle
    b = dt.datetime(2022,10,12,0,0)

    print('<<<<<<<<<Scenarios>>>>>>>>>',tp.get_scenarios_by_time(b))

```

File: taipy/core/taipy.py [hunk 4]
```diff
@@ -1038,3 +1038,23 @@ def get_entities_by_config_id(
     if entities := _DataManagerFactory._build_manager()._get_by_config_id(config_id):
         return entities
     return entities
+
+def get_scenarios_by_time(at: datetime) -> List[Scenario]:
+    """Retrieve the all scenarios occurring within a specific period.
```
Comment (jrobinAV) [inline taipy/core/taipy.py:1043]:
```suggestion
    """Retrieve all scenarios whose cycle period includes a specific date.
```

File: taipy/core/taipy.py [hunk 5]
```diff
@@ -1038,3 +1038,23 @@ def get_entities_by_config_id(
...
+
+    Parameters:
+        at (datetime): The date on which we want to get it's occurring scenarios.
+
+    Returns:
+        The scenarios  of the given time. If the time has no
+            scenario, this method returns an empty List.
```
Comment (jrobinAV) [inline taipy/core/taipy.py:1050]:
```suggestion
        The scenarios whose cycle period includes the given date. If the date does not
            match any scenario, the method returns an empty List.
```

File: tests/core/cycle/test_cycle_manager.py [hunk 6]
```diff
@@ -290,3 +290,20 @@ def test_get_primary(tmpdir, cycle, current_datetime):
...
+    creation_date_2 = datetime(2024, 1, 2, 1, 0, 0, 0)
+    scenario_config_2 = Config.configure_scenario("sc_2", None, None, Frequency.YEARLY)
+    scenario_2 = _ScenarioManager._create(scenario_config_2, creation_date=creation_date_2, name="name_2")
+
+    _ScenarioManager._set(scenario)
+    _ScenarioManager._set(scenario_2)
+
+
+    assert len(_CycleManager._get_cycles_by_overlapping_date(creation_date_1)) == 2
+    assert len(_CycleManager._get_cycles_by_overlapping_date(datetime(2024, 1, 5, 1, 0, 0, 0))) == 2
```
Comment (jrobinAV) [inline tests/core/cycle/test_cycle_manager.py:309]:
```suggestion
    assert len(_CycleManager._get_cycles_by_overlapping_date(datetime(2024, 1, 5, 1, 0, 0, 0))) == 2
    assert len(_CycleManager._get_cycles_by_overlapping_date(datetime(2024, 2, 1, 1, 0, 0, 0))) == 1
    assert len(_CycleManager._get_cycles_by_overlapping_date(datetime(2025, 1, 1, 1, 0, 0, 0))) == 0
    
```

File: taipy/core/taipy.py [hunk 7]
```diff
@@ -1038,3 +1038,23 @@ def get_entities_by_config_id(
     if entities := _DataManagerFactory._build_manager()._get_by_config_id(config_id):
         return entities
     return entities
+
+def get_scenarios_by_time(at: datetime) -> List[Scenario]:
+    """Retrieve the all scenarios occurring within a specific period.
+
+    Parameters:
+        at (datetime): The date on which we want to get it's occurring scenarios.
```
Comment (jrobinAV) [inline taipy/core/taipy.py:1046]:
```suggestion
        at (datetime): The specific date to match the cycle period.
```

**Comment by trgiangdo** [comment]
Hello @Forchapeatl, the `get_scenarios_by_time()` method here is not what discussed in issue https://github.com/Avaiga/taipy/issues/413.

I believe you are on the right track and to ensure alignment with our objectives, I recommend revisiting the specifics of the issue and refining the `get_scenarios()` method accordingly

**Comment by Forchapeatl** [comment]
@trgiangdo , you are right . This PR no-longer addresses  issue #413. I am just holding on to [this comment](https://github.com/Avaiga/taipy/pull/646#pullrequestreview-1804400574)  .

**Comment by trgiangdo** [comment]
Yes, I believe that comment is obsolete since we were in the discussion phase. Can you help us implement the API as discussed in the issue? I believe we can use a lot of the code you already implemented

**Comment by trgiangdo** [comment]
Closed due to inactivity

**Pull request closed without merging**